cpp/src/arrow/compute/kernels/codegen_internal.h

```diff
@@ -455,6 +458,41 @@ static void VisitTwoArrayValuesInline(const ArraySpan& arr0, const ArraySpan& ar
                         std::move(visit_null));
 }
 
+template <typename Arg0Type, typename Arg1Type, typename VisitBothNotNull,
+          typename VisitLeftNull, typename VisitRightNull, typename VisitBothNull>
+static void VisitTwoArrayValuesInlineAllCases(const ArraySpan& arr0,
+                                              const ArraySpan& arr1,
+                                              VisitBothNotNull&& visit_both_not_null,
+                                              VisitLeftNull&& visit_left_null,
+                                              VisitRightNull&& visit_right_null,
+                                              VisitBothNull&& visit_both_null) {
+  ArrayIterator<Arg0Type> arr0_it(arr0);
+  ArrayIterator<Arg1Type> arr1_it(arr1);
+
+  auto forward_visit_both_not_null = [&](int64_t i) {
```

Comment:
Maybe `[&]` should be replaced by `[visit_both_not_null = std::move(visit_both_not_null)]` so the r-value parameters are moved into the local lambdas. As they are not shared, they can be moved.

Reply:
`arr0_it` and `arr1_it` would be captured by `&` of course as they are shared among the different lambdas.

Reply:
I think capturing by reference is better here because the lambda won't need to create a variable for the func. See https://cppinsights.io/s/690fcf31 for an example.